gui/src/app/RunPanel/RunPanel.tsx

```diff
@@ -39,62 +45,94 @@ const RunPanel: FunctionComponent<RunPanelProps> = ({
-        <div>
+
+  let content;
+  if (compileStatus === "compiled") {
+    content = (
+      <div>
+        <Button
+          variant="contained"
+          color="success"
+          onClick={handleRun}
+          disabled={runStatus === "sampling" || runStatus === "loading"}
+        >
+          run sampling
+        </Button>
+        &nbsp;
+        {runStatus === "sampling" && (
           <Button
-            variant="contained"
-            color="success"
-            onClick={handleRun}
-            disabled={runStatus === "sampling" || runStatus === "loading"}
+            color="error"
+            variant="outlined"
+            onClick={cancelRun}
+            disabled={runStatus !== "sampling"}
           >
-            run sampling
+            cancel
           </Button>
-          &nbsp;
-          {runStatus === "sampling" && (
-            <Button
-              color="error"
-              variant="outlined"
-              onClick={cancelRun}
-              disabled={runStatus !== "sampling"}
-            >
-              cancel
-            </Button>
-          )}
-          <hr />
-          {runStatus === "loading" && <div>Loading compiled Stan model...</div>}
-          {runStatus === "sampling" && (
-            <div>
-              Sampling
-              <SamplingProgressComponent
-                report={progress}
-                numChains={samplingOpts.num_chains}
-              />
-            </div>
-          )}
-          {runStatus === "completed" && <div>done sampling</div>}
-          {runStatus === "failed" && (
-            <div>
-              Sampling failed!
-              <pre className="SamplerError">{errorMessage}</pre>
-              <span className="details">
-                (see browser console for more details)
-              </span>
-            </div>
-          )}
-        </div>
+        )}
+        <hr />
+        {runStatus === "loading" && <div>Loading compiled Stan model...</div>}
+        {runStatus === "sampling" && (
+          <div>
+            Sampling
+            <SamplingProgressComponent
+              report={progress}
+              numChains={samplingOpts.num_chains}
+            />
+          </div>
+        )}
+        {runStatus === "completed" && <div>done sampling</div>}
+        {runStatus === "failed" && (
+          <div>
+            Sampling failed!
+            <pre className="SamplerError">{errorMessage}</pre>
+            <span className="details">
+              (see browser console for more details)
+            </span>
+          </div>
+        )}
+      </div>
+    );
+  } else if (["preparing", "compiling"].includes(compileStatus)) {
+    content = <div>{compileMessage}</div>;
+  } else {
+    content = (
+      <div>
+        <Button
+          variant="contained"
+          onClick={compile}
+          disabled={isCompileModelDisabled(projectData, validSyntax)}
+        >
+          compile model
+        </Button>
       </div>
+    );
+  }
```

Comment:
Even pulled out of the return statement itself, this is still pretty long & tough to follow; the branch where the Stan code has successfully been compiled, in particular, is quite a mouthful.

I think I'd be happier if the 3 branches of the conditional were placed in components & then the conditional just chose from the 3 components.

Actually, I think I'd want to put the entire `compileStatus==="compiled"` branch in a separate file, and have that return 3-4 individual versions based on the compile status--but having the conditional at the top rather than running throughout the single `div` element.

Let me know if you'd prefer that I just do this--I know it can be frustrating to have someone ask for something fairly specific without giving enough details.

Reply:
Sure, if you want to do that would be helpful. If I understand what you are proposing, I agree with it.

Reply:
If you don't mind waiting, give me 10-15 and I'll put in a PR against this branch.